ruma-common/src/push/condition.rs

```diff
@@ -42,13 +52,273 @@ pub enum PushCondition {
+    /// a glob with wildcards `*` and `?`.
+    fn matches_pattern(&self, pattern: &str, match_words: bool) -> bool;
+
+    /// Matches the given value against a pattern, with word boundaries.
+    ///
+    /// The match is case sensitive.
+    fn matches_word(&self, pattern: &str) -> bool;
+}
+
+impl StrExt for str {
+    fn char_len(&self, index: usize) -> usize {
+        let mut len = 1;
+        while !self.is_char_boundary(index + len) {
+            len += 1;
+        }
+        len
+    }
+
+    fn char_at(&self, index: usize) -> char {
+        let end = index + self.char_len(index);
+        let char_str = &self[index..end];
+        char::from_str(char_str)
+            .unwrap_or_else(|_| panic!("Could not convert str '{}' to char", char_str))
+    }
+
+    fn find_prev_char(&self, index: usize) -> Option<char> {
+        if index == 0 {
+            return None;
+        }
+
+        let mut pos = index - 1;
+        while !self.is_char_boundary(pos) {
+            pos -= 1;
+        }
+        Some(self.char_at(pos))
+    }
+
+    fn matches_pattern(&self, pattern: &str, match_words: bool) -> bool {
+        if self.is_empty() || pattern.is_empty() {
+            return false;
+        }
+
+        let value = &self.to_lowercase();
+        let pattern = &pattern.to_lowercase();
+
+        if match_words {
+            value.matches_word(pattern)
+        } else {
+            WildMatch::new(pattern).matches(value)
+        }
+    }
+
+    fn matches_word(&self, pattern: &str) -> bool {
+        match self.find(pattern) {
+            Some(start) => {
+                let end = start + pattern.len();
+
+                // Look if the match has word boundaries.
+                let word_boundary_start = if !self.char_at(start).is_word_char() {
+                    true
+                } else {
+                    match self.find_prev_char(start) {
+                        Some(c) => !c.is_word_char(),
+                        None => true,
+                    }
+                };
+
+                if word_boundary_start {
+                    let word_boundary_end = end == self.len()
+                        || !self.find_prev_char(end).unwrap().is_word_char()
+                        || !self.char_at(end).is_word_char();
+
+                    if word_boundary_end {
+                        return true;
+                    }
+                }
+
+                // Find next word.
+                let non_word_str = &self[start..];
+                let non_word = match non_word_str.find(|c: char| !c.is_word_char()) {
+                    Some(pos) => pos,
+                    None => return false,
+                };
+
+                let word_str = &non_word_str[non_word..];
+                let word = match word_str.find(|c: char| c.is_word_char()) {
+                    Some(pos) => pos,
+                    None => return false,
+                };
+
+                (&word_str[word..]).matches_word(pattern)
```

Comment:
I think this should work due to autoref:

```suggestion
                word_str[word..].matches_word(pattern)
```

Reply:
It does work!